PYIODE: fix generating stub files *.pyi on Github Actions

See for example https://github.com/plan-be/iode/actions/runs/10469077164/job/28991292222?pr=640 
```bash
Run python generate_stub_files.py
Traceback (most recent call last):
stubgen --package iode --include-docstrings --output D:/a/iode/iode/pyiode --verbose
Trying to import 'iode' for runtime introspection
Trying to import 'iode' for runtime introspection
Trying to import 'iode.deprecated' for runtime introspection
  File "D:\a\iode\iode\generate_stub_files.py", line 22, in <module>
    main()
  File "mypy\stubgen.py", line 1895, in main
Trying to import 'iode.deprecated.compat_larray' for runtime introspection
Trying to import 'iode.deprecated.compat_pandas' for runtime introspection
Trying to import 'iode.deprecated.estimation' for runtime introspection
Trying to import 'iode.deprecated.iode_objects' for runtime introspection
Trying to import 'iode.deprecated.lec' for runtime introspection
Trying to import 'iode.deprecated.model' for runtime introspection
Trying to import 'iode.deprecated.report' for runtime introspection
Trying to import 'iode.deprecated.time' for runtime introspection
Trying to import 'iode.deprecated.workspace' for runtime introspection
Trying to import 'iode.iode_python' for runtime introspection
Trying to import 'iode.util' for runtime introspection
Processing 12 files...
  File "mypy\stubgen.py", line 1691, in generate_stubs
  File "mypy\stubgen.py", line 1623, in generate_asts_for_modules
  File "mypy\build.py", line 193, in build
  File "mypy\build.py", line 268, in _build
  File "mypy\build.py", line 2950, in dispatch
  File "mypy\build.py", line 3348, in process_graph
  File "mypy\build.py", line 3475, in process_stale_scc
  File "mypy\build.py", line 2507, in write_cache
  File "mypy\build.py", line 1568, in write_cache
  File "mypy\nodes.py", line 390, in serialize
  File "mypy\nodes.py", line 4012, in serialize
  File "mypy\nodes.py", line 3949, in serialize
  File "mypy\nodes.py", line 3374, in serialize
  File "mypy\types.py", line 671, in serialize
  File "mypy\types.py", line 2428, in serialize
  File "mypy\types.py", line 1451, in serialize
  File "mypy\types.py", line 671, in serialize
  File "mypy\types.py", line 3093, in serialize
AssertionError: Internal error: unresolved placeholder type None
Error: Process completed with exit code 1.
```

--- a/.github/workflows/github-actions-python.yml
+++ b/.github/workflows/github-actions-python.yml
@@ -80,13 +80,14 @@ jobs:
           - name: Check conda
             shell: bash -el {0}
             run: conda info
-          - name: Install Python libraires - numpy, pandas, larray, cython, scikit-build-core and pytest     # nanobind
+          - name: Install Python libraires
             shell: bash -el {0}
+            # fixed numpy<2.0 since newer versions make mypy to crash 
+            # (see mypy issue https://github.com/python/mypy/issues/17538)
             run: |
               conda config --add channels conda-forge
               conda config --add channels larray-project
-              conda install numpy pandas larray cython scikit-build-core pytest
-            # conda install nanobind
+              conda install "numpy<2.0" pandas larray cython mypy scikit-build-core pytest
           - name: Configure CMake
             shell: bash -el {0}
             # cmake --preset <configurePreset> where <configurePreset> is the name of the active Configure Preset.
@@ -95,6 +96,9 @@ jobs:
             #   - https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html
             #   - https://docs.microsoft.com/en-us/cpp/build/cmake-presets-vs?view=msvc-170
             run: cmake --preset ${{ env.os }}-${{ inputs.build-config }}
+          - name: Build target iode_python before pip wheel to make sure it compiles
+            shell: bash -el {0}
+            run: cmake --build --preset ${{ env.os }}-${{ inputs.build-config }} --target iode_python
           - name: Build iode wheel file
             shell: bash -el {0}
             run: pip wheel .

--- a/CONTRIBUTE.md
+++ b/CONTRIBUTE.md
@@ -173,18 +173,10 @@ root_dir_iode> pip install <iode-wheel-file>.whl
 To help IDE like PyCharm to understand the Python API of `iode`, 
 stub files *.pyi must be generated and added to installed package.
 
-To generate the stub files, you must first build and install the Python `iode` package:
-```bash
-root_dir_iode> pip wheel .
-root_dir_iode> pip install <iode-wheel-file>.whl
-```
-Then, type the following command in a console: 
-```bash
-root_dir_iode> python generate_stub_files.py
-```
-This command will generate the stub files in the directory `pyiode/iode`.
+The stub files are generated using the tool `stubgen` from the package `mypy`.
+They are automatically generated when the CMake target `iode_python` is built.
 
-Finally, commit the updated stub files to the repository.
+See pyiode/CMakeLists.txt and generate_stub_files.py for more details.
 
 # GUI (Python)
 Before to work on any issue related to the GUI (Python), you have to build the CMake target `ui_to_py`. 

--- a/doc/requirements.txt
+++ b/doc/requirements.txt
@@ -1,5 +1,5 @@
 # dependencies to build the iode module (via nbsphinx)
-numpy 
+numpy < 2.0     # numpy >= 2.0 makes mypy (1.11) to crash (21/08/2024)
 pandas 
 matplotlib
 larray
@@ -7,6 +7,7 @@ larray
 cmake
 cython
 scikit-build-core
+mypy
 
 # dependencies to actually build the documentation
 sphinx

--- a/generate_stub_files.py.in
+++ b/generate_stub_files.py.in
@@ -2,6 +2,7 @@ import re
 import sys
 import inspect
 import mypy.stubdoc
+from pathlib import Path
 from typing import Optional
 
 
@@ -41,7 +42,8 @@ import mypy.stubgen
 if __name__ == "__main__":
     # NOTE: see https://mypy.readthedocs.io/en/latest/stubgen.html 
     #       for options of stubgen command
-    sys.argv = ["stubgen", "--package", "iode", "--include-docstrings", "--output", 
-                "${CMAKE_CURRENT_SOURCE_DIR}", "--verbose"]
+    sys.argv = ["stubgen", "--package", "iode", "--include-docstrings", 
+                "--search-path", "${CMAKE_CURRENT_BINARY_DIR}", 
+                "--output", "${CMAKE_CURRENT_SOURCE_DIR}", "--verbose"]
     print(' '.join(sys.argv))
     mypy.stubgen.main()

--- a/pyiode/CMakeLists.txt
+++ b/pyiode/CMakeLists.txt
@@ -20,6 +20,9 @@ if(Python_FOUND)
 
     file(INSTALL ${CMAKE_SOURCE_DIR}/iode.msg DESTINATION ${CMAKE_CURRENT_SOURCE_DIR}/iode)
     file(INSTALL ${CMAKE_SOURCE_DIR}/tests/data DESTINATION ${CMAKE_CURRENT_SOURCE_DIR}/iode/tests)
+    # copy all Python files to binary directory to then be able to generate stub files *.pyi
+    file(INSTALL ${CMAKE_CURRENT_SOURCE_DIR}/iode DESTINATION ${CMAKE_CURRENT_BINARY_DIR} 
+         FILES_MATCHING PATTERN "*.py")
 
     if(EXISTS "${CMAKE_CURRENT_BINARY_DIR}/iode_python.cpp")
       file(REMOVE "${CMAKE_CURRENT_BINARY_DIR}/iode_python.cpp")
@@ -60,6 +63,12 @@ if(Python_FOUND)
     target_include_directories(iode_python PRIVATE ${CMAKE_SOURCE_DIR}/scr4 ${CMAKE_SOURCE_DIR}/api ${CMAKE_SOURCE_DIR}/cpp_api)
     target_compile_definitions(iode_python PUBLIC NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION)
 
+    # Add a custom command that runs the Python script to generate stub files 
+    # after the iode_python target is built
+    add_custom_command(TARGET iode_python POST_BUILD
+        COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:iode_python> ${CMAKE_CURRENT_BINARY_DIR}/iode
+        COMMAND ${Python_EXECUTABLE} ${CMAKE_SOURCE_DIR}/generate_stub_files.py)
+    
     install(TARGETS iode_python LIBRARY DESTINATION iode)
   else()
     message(CHECK_FAIL "Cython not found")

--- a/pyproject.toml.in
+++ b/pyproject.toml.in
@@ -1,7 +1,7 @@
 # configuration file for the iode python package
 
 [build-system]
-requires = [ "scikit-build-core >=0.4.3", "numpy", "cython>=3.0" ]
+requires = [ "scikit-build-core >=0.4.3", "numpy", "cython>=3.0", "mypy>=1.10" ]
 build-backend = "scikit_build_core.build"
 
 [project]

